### ISSUE TITLE
How should ArgumentParser metadata be provided?

### ISSUE THREAD

**Issue by arezaii** [chapel-lang/chapel#18648]
Continuing work on the `ArgumentParser` library, we would like to improve our generated help message output by including additional fields, such as a program's name, description, author, version, and usage text. 

Python's `argparse` library allows a user to define some of these values during the instantiation of the `ArgumentParser` by using parameters such as:

- `prog` - to set the name of the program (defaults to sys.argv[0])
- `usage` - string describing program usage (defaults to string generated from the arguments added)
- `description` - text to display before the help message (default none)
- `epilog` - text to display after the help message (default none)

Similarly, Rust's `clap` crate defines an expanded list of methods on the `App` struct:
- `name` - program name (typically set when new'ing up an `App` struct)
- `about` - a string describing what the program does
- `author` - program author
- `version` - that one explains itself
- `before_help` - adds additional help information to be displayed in before the auto-generated help
- `help_message` - sets the help text for the auto-generated `help` flag
- `after_help` - adds additional help information to be displayed in after the auto-generated help
- `usage` - a custom usage string to override the auto-generated usage string
- `bin_name` - overrides the system-determined binary name

Swift's `ArgumentParser` defines a multi-purpose `CommandConfiguration` struct that contains some of this information as well (here `Command` may be interpreted as `program`), such as:

- `commandName` - the name of the command to use on the command line
- `abstract` - a one-line description of the command
- `discussion` - a longer description of the command
- `version` - version number for the command
- `helpNames` - overrides default flags for help (`-h`/`--help`)

I would propose adding parameters with similar names to Rust's `clap` to the `init` method of our `argumentParser` record to capture this metadata.

An implementation might look like:
```chapel
proc init(<previously defined parameters>, name="", 
                author="", version="", beforeHelp="", afterHelp="",
                usage="", helpFlags=["-h","--help"], helpMessage="Prints help information")
```
where additional fields that are defaulted to the empty string indicate they are not applicable or should be auto-generated as in the case of `name` and `usage`.

Rust's `clap` defines many separate values for help message outputs that are displayed differently depending on whether the long version (`--help`) or short version (`-h`) of the `help` flag was detected, but I am not proposing that we add similar functionality at this time.

**Comment by dlongnecke-cray** [comment]
I like the look of the `Rust` formals myself. Seems to be the most configurable and the names read better IMO than the ones from Swift and Python.

**Comment by aconsroe-hpe** [comment]
Is using the empty string as a standin for `None` going to lead to surprising results? If I wanted to print no usage, passing `usage=""` doesn't do that. I might resort to `usage=" "`, which feels a bit hacky. Not able to immediately think of an alternative.

**Comment by arezaii** [comment]
> If I wanted to print no usage, passing `usage=""` doesn't do that

Yes, that would be a possible downside, but should we really be disabling the usage message entirely? I can see replacing it with something, but I'm not sure I see the use case for displaying _nothing_ in place of the generated message. What kind of monster would do that to their users? 🐱  (no offense to the monsters)

**Comment by dlongnecke-cray** [comment]
I think if we decide we wanted to completely disable some features, it would be reasonable to add a second, more verbose initializer at a later point in time if possible (?).

**Comment by mppf** [comment]
Can the version integrate somehow with the version module https://chapel-lang.org/docs/modules/standard/Version.html ? I think we have been hoping that modules use a global variable or something to declare the current module version.

In the Rust approach, are these actually provided by defining methods? Rather than as arguments? I think I find the Swift approach the most appealing to emulate of the 3, myself. But, I don't have a strong opinion about this.

**Comment by arezaii** [comment]
> Can the version integrate somehow with the version module ...

Is the idea to have the argument parser store a global variable that holds the program's version? Or just have the argument parser use a `sourceVersion` as opposed to a string?
My initial concern is that it would force everyone who uses the argument parser to use the same version format that `sourceVersion` defines, regardless of if they intended to build a mason package or not.


> ... are these actually provided by defining methods?

I may not be using the proper terminology for Rust. I believe these are methods in Rust: https://docs.rs/clap/2.33.3/clap/struct.App.html#method.usage , for example.

**Comment by mppf** [comment]
> Or just have the argument parser use a sourceVersion as opposed to a string?

That's what I was thinking

> I believe these are methods in Rust

Oh, I see, they are effectively setters

**Comment by dlongnecke-cray** [comment]
I like the idea of subclassing more than a config struct, as it feels more high level (whereas I associate the config struct with a low level language like C). I also bet you can produce some more powerful behavior.

**Comment by bradcray** [comment]
> I like the idea of subclassing more than a config struct, as it feels more high level (whereas I associate the config struct with a low level language like C). I also bet you can produce some more powerful behavior.

David, I'm not understanding your comment.  What do you mean by a `config struct`?  In what way is subclassing in play on this issue?  (I'm not finding either term mentioned before your comment searching this issue, so feel like there's context I'm missing).

**Comment by dlongnecke-cray** [comment]
I interpreted this line in the OP:

> Swift's ArgumentParser defines a multi-purpose CommandConfiguration struct that contains some of this information as well (here Command may be interpreted as program), such as:

To mean that the Swift argument parser was configured by passing a struct/record as an argument. That plus some of Michael's comments, though I certainly could have interpreted things wrong.

I might have also gotten this mixed up with: https://github.com/chapel-lang/chapel/issues/18646

:D
